File: compiler/src/test/java/motif/compiler/TestHarness.java
/*
 * Copyright (c) 2018-2019 Uber Technologies, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package motif.compiler;

import com.google.common.io.Files;
import com.tschuchort.compiletesting.KotlinCompilation;
import com.tschuchort.compiletesting.SourceFile;
import dagger.internal.codegen.ComponentProcessor;
import kotlin.text.StringsKt;
import motif.core.ResolvedGraph;
import motif.errormessage.ErrorMessage;
import motif.viewmodel.TestRenderer;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import javax.annotation.Nullable;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.charset.Charset;
import java.nio.file.Path;
import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static com.google.common.truth.Truth.assertWithMessage;

@RunWith(Parameterized.class)
public class TestHarness {

    private static final File SOURCE_ROOT = new File("../tests/src/main/java/");
    private static final File TEST_CASE_ROOT = new File(SOURCE_ROOT, "testcases");
    private static final File EXTERNAL_ROOT = new File(SOURCE_ROOT, "external");

    private static class TestParameters {

        final File testDirectory;
        final Mode mode;

        public TestParameters(File testDirectory, Mode mode) {
            this.testDirectory = testDirectory;
            this.mode = mode;
        }
    }

    @Parameterized.Parameters(name = "{1}")
    public static Collection<Object[]> data() {
        File[] testCaseDirs = TEST_CASE_ROOT.listFiles(TestHarness::isTestDir);
        if (testCaseDirs == null) throw new IllegalStateException("Could not find test case directories: " + TEST_CASE_ROOT);
        return Arrays.stream(Mode.values())
                .flatMap(mode -> Arrays.stream(testCaseDirs).map(file -> new TestParameters(file, mode)))
                .filter(parameters -> !skipTest(parameters.mode, parameters.testDirectory))
                .map(parameters -> {
                    String displayName = parameters.testDirectory.getName() + "_" + parameters.mode;
                    return new Object[]{parameters, displayName};
                })
                .collect(Collectors.toList());
    }

    private static boolean skipTest(Mode mode, File testDirectory) {
        if (mode == Mode.KOTLIN && skipKotlin(testDirectory)) {
            return true;
        }
        return false;
    }

    private static boolean skipKotlin(File testDirectory) {
        return new File(testDirectory, "SKIP_KOTLIN").exists();
    }

    private static boolean isTestDir(File file) {
        String filename = file.getName();
        return file.isDirectory()
                && file.listFiles().length > 0
                && (filename.startsWith("T") || filename.startsWith("E"));
    }

    @Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();

    private final Mode mode;
    private final File testCaseDir;
    private final File externalDir;
    private final String testClassName;
    private final File errorFile;
    private final File graphFile;
    private final File proguardFile;
    private final boolean isErrorTest;

    @SuppressWarnings("unused")
    public TestHarness(TestParameters parameters, String displayName) {
        String testName = parameters.testDirectory.getName();
        this.mode = parameters.mode;
        this.testCaseDir = new File(TEST_CASE_ROOT, testName);
        this.externalDir = new File(EXTERNAL_ROOT, testName);
        this.testClassName = "testcases." + testName + ".Test";
        this.errorFile = new File(testCaseDir, "ERROR.txt");
        this.graphFile = new File(testCaseDir, "GRAPH.txt");
        this.proguardFile = new File(testCaseDir, "config.pro");
        this.isErrorTest = testName.startsWith("E");
    }

    @Test
    public void test() throws Throwable {
        Map<String, String> options = new HashMap<>();
        options.put("motif.mode", mode.name().toLowerCase());

        File externalClassesDir = null;
        File[] externalDirContents = externalDir.listFiles();
        if (externalDirContents != null && externalDirContents.length > 0) {
            boolean shouldProcess = !new File(externalDir, "DO_NOT_PROCESS").exists();
            KotlinCompilation.Result externalResult = compile(
                    getFiles(externalDir),
                    Collections.emptyList(),
                    options,
                    shouldProcess ? new Processor() : null);
            assertSucceeded(externalResult);
            externalClassesDir = externalResult.getOutputDirectory();
        }

        List<File> classpaths = externalClassesDir == null
                ? Collections.emptyList()
                : Collections.singletonList(externalClassesDir);
        Processor processor = new Processor();

        KotlinCompilation.Result result = compile(
                getFiles(testCaseDir),
                classpaths,
                options,
                processor);

        if (isErrorTest) {
            runErrorTest(result);
        } else {
            assertSucceeded(result);

            File proguardedClasses = ProGuard.run(
                    externalClassesDir,
                    result.getOutputDirectory(),
                    proguardFile);

            ClassLoader classLoader;
            if (externalClassesDir == null) {
                classLoader = new URLClassLoader(
                        new URL[]{
                                proguardedClasses.toURI().toURL()
                        },
                        getClass().getClassLoader());
            } else {
                classLoader = new URLClassLoader(
                        new URL[]{
                                proguardedClasses.toURI().toURL(),
                                externalClassesDir.toURI().toURL()
                        }, getClass().getClassLoader());
            }
            Class<?> testClass = classLoader.loadClass(testClassName);

            if (testClass.getAnnotation(Ignore.class) == null) {
                runSuccessTest(testClass, processor.graph);
            }
        }
    }

    private void assertSucceeded(KotlinCompilation.Result result) {
        if (result.getExitCode() != KotlinCompilation.ExitCode.OK) {
            assertWithMessage(result.getMessages()).fail();
        }
    }

    private void assertFailed(KotlinCompilation.Result result) {
        if (result.getExitCode() == KotlinCompilation.ExitCode.OK) {
            assertWithMessage("Expected compilation to fail but encountered no errors.").fail();
        }
    }

    private KotlinCompilation.Result compile(
            List<File> sourceFiles,
            List<File> classpaths,
            Map<String, String> aptArgs,
            @Nullable Processor motifProcessor) {
        KotlinCompilation compilation = new KotlinCompilation();
        compilation.setAnnotationProcessors(getProcessors(motifProcessor));
        compilation.setSources(sourceFiles.stream()
                .map(SourceFile.Companion::fromPath)
                .collect(Collectors.toList()));
        compilation.setInheritClassPath(true);
        compilation.setClasspaths(classpaths);
        compilation.setKaptArgs(aptArgs);
        compilation.setVerbose(false);

        return compilation.compile();
    }

    private List<File> getFiles(File dir) throws IOException {
        return getFiles(dir, file -> (file.getName().endsWith(".java") || file.getName().endsWith(".kt")) && !file.getName().equals("ScopeImpl.java"));
    }

    private List<File> getFiles(File dir, Predicate<File> filter) throws IOException {
        return java.nio.file.Files.walk(dir.toPath())
                .map(Path::toFile)
                .filter(file -> !file.isDirectory() && filter.test(file))
                .collect(Collectors.toList());
    }

    private List<javax.annotation.processing.Processor> getProcessors(@Nullable Processor motifProcessor) {
        List<javax.annotation.processing.Processor> processors = new ArrayList<>();
        processors.add(new ComponentProcessor());
        if (motifProcessor != null) {
            processors.add(motifProcessor);
        }
        return processors;
    }

    private void runErrorTest(KotlinCompilation.Result result) throws IOException {
        assertFailed(result);

        String expectedErrorString = getExistingErrorString();
        String actualErrorString = getActualErrorString(result);

        if (!expectedErrorString.equals(actualErrorString)) {
            try (BufferedWriter out = new BufferedWriter(new FileWriter(errorFile))) {
                out.write(actualErrorString);
            }
            assertWithMessage("Error message has changed. The ERROR.txt file has been " +
                    "automatically updated by this test:\n" +
                    "  1. Verify that the changes are correct.\n" +
                    "  2. Commit the changes to source control.\n").fail();
        }
    }

    private void runSuccessTest(Class<?> testClass, ResolvedGraph graph) throws Throwable {
        try {
            testClass.getMethod("run").invoke(null);
        } catch (InvocationTargetException e) {
            throw e.getCause();
        }

        String expectedGraphString = getExistingGraphString();
        String actualGraphString = getActualGraphString(graph);

        if (!expectedGraphString.equals(actualGraphString)) {
            try (BufferedWriter out = new BufferedWriter(new FileWriter(graphFile))) {
                out.write(actualGraphString);
            }
            assertWithMessage("Graph representation has changed. The GRAPH.txt file has been " +
                    "automatically updated by this test:\n" +
                    "  1. Verify that the changes are correct.\n" +
                    "  2. Commit the changes to source control.\n").fail();
        }
    }

    private String getActualGraphString(ResolvedGraph graph) {
        String message = TestRenderer.render(graph);
        String header =
                "########################################################################\n" +
                "#                                                                      #\n" +
                "# This file is auto-generated by running the Motif compiler tests and  #\n" +
                "# serves a as validation of graph correctness. IntelliJ plugin tests   #\n" +
                "# also rely on this file to ensure that the plugin graph understanding #\n" +
                "# is equivalent to the compiler's.                                     #\n" +
                "#                                                                      #\n" +
                "# - Do not edit manually.                                              #\n" +
                "# - Commit changes to source control.                                  #\n" +
                "# - Since this file is autogenerated, code review changes carefully to #\n" +
                "#   ensure correctness.                                                #\n" +
                "#                                                                      #\n" +
                "########################################################################\n";
        return header + "\n" + message + "\n";
    }

    private String getExistingGraphString() throws IOException {
        if (graphFile.exists()) {
            return Files.asCharSource(graphFile, Charset.defaultCharset()).read();
        } else {
            return "";
        }
    }

    private String getActualErrorString(KotlinCompilation.Result result) {
        String message = getMessage(result);
        String header =
                "########################################################################\n" +
                "#                                                                      #\n" +
                "# This file is auto-generated by running the Motif compiler tests and  #\n" +
                "# serves both as validation of error correctness and as a record of    #\n" +
                "# the current compiler error output.                                   #\n" +
                "#                                                                      #\n" +
                "# - Do not edit manually.                                              #\n" +
                "# - Commit changes to source control.                                  #\n" +
                "# - Since this file is autogenerated, code review changes carefully to #\n" +
                "#   ensure correctness.                                                #\n" +
                "#                                                                      #\n" +
                "########################################################################\n";
        return header + message + "\n";
    }

    private String getMessage(KotlinCompilation.Result result) {
        String resultMessage = result.getMessages();
        String header = toCompilerMessage(ErrorMessage.Companion.getHeader());
        String footer = toCompilerMessage(ErrorMessage.Companion.getFooter());
        resultMessage = StringsKt.substringAfter(resultMessage, header, resultMessage);
        resultMessage = StringsKt.substringBefore(resultMessage, footer, resultMessage);
        return "\n" + header + resultMessage + footer;
    }

    private String toCompilerMessage(String message) {
        return StringsKt.prependIndent(message.trim(), "  ");
    }

    private String getExistingErrorString() throws IOException {
        if (errorFile.exists()) {
            return Files.asCharSource(errorFile, Charset.defaultCharset()).read();
        } else {
            return "";
        }
    }
}

File: compiler/src/test/java/motif/compiler/TestHarness.kt
/*
 * Copyright (c) 2022 Uber Technologies, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package motif.compiler

import androidx.room.compiler.processing.ExperimentalProcessingApi
import androidx.room.compiler.processing.XProcessingEnvConfig
import androidx.room.compiler.processing.util.Source
import androidx.room.compiler.processing.util.compiler.TestCompilationArguments
import androidx.room.compiler.processing.util.compiler.TestCompilationResult
import androidx.room.compiler.processing.util.compiler.compile
import com.google.common.collect.Lists.cartesianProduct
import com.google.common.truth.Truth
import com.google.devtools.ksp.processing.SymbolProcessorProvider
import dagger.internal.codegen.ComponentProcessor
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.net.URLClassLoader
import java.nio.charset.Charset
import java.nio.file.Files
import javax.tools.Diagnostic
import motif.compiler.ksp.MotifSymbolProcessorProvider
import motif.core.ResolvedGraph
import motif.errormessage.ErrorMessage.Companion.footer
import motif.errormessage.ErrorMessage.Companion.header
import motif.viewmodel.TestRenderer.render
import org.junit.Ignore
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

@RunWith(Parameterized::class)
@ExperimentalProcessingApi
class TestHarness(
    private val processorType: ProcessorType,
    private val outputMode: OutputMode,
    private val testCaseDir: File,
    testName: String,
) {

  @Rule @JvmField var temporaryFolder = TemporaryFolder()

  private val externalDir = EXTERNAL_ROOT.resolve(testName)
  private val testClassName = "testcases.$testName.Test"
  private val errorFile = testCaseDir.resolve("ERROR.txt")
  private val graphFile = testCaseDir.resolve("GRAPH.txt")
  private val proguardFile = testCaseDir.resolve("config.pro")
  private val isErrorTest = testName.matches("^K?E.*".toRegex())

  companion object {
    private val SOURCE_ROOT = File("../tests/src/main/java/")
    private val TEST_CASE_ROOT = File(SOURCE_ROOT, "testcases")
    private val EXTERNAL_ROOT = File(SOURCE_ROOT, "external")

    // Used for debugging. Run all by default
    private val testFilter = { proc: ProcessorType, mode: OutputMode, dir: File, name: String ->
      true
    }

    @JvmStatic
    @Parameterized.Parameters(name = "{0}_{1}_{3}")
    fun data(): Collection<Array<Any>> {
      val testCaseDirs = TEST_CASE_ROOT.listFiles { file: File -> isTestDir(file) }
      val combos =
          cartesianProduct(
              ProcessorType.values().toList(), OutputMode.values().toList(), testCaseDirs.toList())
      return combos
          .filterNot { (_, mode, dir) ->
            mode == OutputMode.KOTLIN && (dir as File).resolve("SKIP_KOTLIN").exists()
          }
          .map { it + (it.last() as File).name }
          .filterNot { (proc, mode, _, _) ->
            // We don't generate Java from KSP
            proc == ProcessorType.KSP && mode == OutputMode.JAVA
          }
          .filterNot { (proc, _, _, name) ->
            // Can't run KSP on Java sources until fixed: https://github.com/google/ksp/issues/1086
            proc == ProcessorType.KSP && "$name"[0] in setOf('T', 'E')
          }
          .filter { (proc, mode, dir, name) ->
            testFilter(proc as ProcessorType, mode as OutputMode, dir as File, name as String)
          }
          .map { it.toTypedArray() as Array<Any> }
    }

    private fun isTestDir(file: File): Boolean {
      return file.isDirectory &&
          file.listFiles().isNotEmpty() &&
          file.name.matches("^K?[TE].*".toRegex())
    }
  }

  @Test
  fun test() {
    val externalClassesDirs = compileExternalDir()
    val (annotationProcessor, symbolProcessorProvider) = createProcessors()
    val result =
        compileDir(testCaseDir, annotationProcessor, symbolProcessorProvider, externalClassesDirs)
    val graph =
        when (processorType) {
          ProcessorType.AP -> annotationProcessor?.graph
          ProcessorType.KSP -> symbolProcessorProvider?.graph
        }
    checkNotNull(graph) { "No graph found during processing" }

    if (isErrorTest) {
      runErrorTest(result, graph)
    } else {
      assertSucceeded(result)

      val proguardedClasses =
          ProGuard.run(
              externalClassesDirs,
              result.outputClasspath.filterNot {
                "/ksp/" in it.absolutePath || "/kapt/" in it.absolutePath
              },
              proguardFile)

      val urls = (externalClassesDirs + proguardedClasses).map { it.toURI().toURL() }.toTypedArray()
      val classLoader: ClassLoader = URLClassLoader(urls, javaClass.classLoader)

      if (externalClassesDirs.isEmpty()) {
        URLClassLoader(arrayOf(proguardedClasses.toURI().toURL()), javaClass.classLoader)
      }
      val testClass = classLoader.loadClass(testClassName)

      if (testClass.getAnnotation(Ignore::class.java) == null) {
        runSuccessTest(testClass, graph)
      }
    }
  }

  private fun compileExternalDir(): List<File> {
    if (externalDir.exists()) {
      val externalDirContents = externalDir.listFiles()
      if (externalDirContents?.isNotEmpty() == true) {
        val (annotationProcessor, symbolProcessorProvider) = createProcessors()
        val shouldProcess = !File(externalDir, "DO_NOT_PROCESS").exists()
        val externalResult =
            compileDir(
                externalDir,
                if (shouldProcess) annotationProcessor else null,
                if (shouldProcess) symbolProcessorProvider else null,
                emptyList())
        assertSucceeded(externalResult)
        return externalResult.outputClasspath.filter { it.listFilesRecursively().isNotEmpty() }
      }
    }
    return emptyList()
  }

  private fun compileDir(
      sourcesDir: File,
      annotationProcessor: javax.annotation.processing.Processor?,
      symbolProcessorProvider: SymbolProcessorProvider?,
      classpath: List<File> = emptyList()
  ): TestCompilationResult {
    val processorOptions = mapOf("motif.mode" to outputMode.name.lowercase())
    val sources = getFiles(sourcesDir).asSources()
    val annotationProcessors =
        annotationProcessor?.let { listOf(it, ComponentProcessor()) } ?: emptyList()
    return compile(
        workingDir = Files.createTempDirectory("test-runner").toFile(),
        arguments =
            TestCompilationArguments(
                sources = sources,
                classpath = classpath,
                inheritClasspath = true,
                kaptProcessors = annotationProcessors,
                symbolProcessorProviders = symbolProcessorProvider?.let { listOf(it) }
                        ?: emptyList(),
                processorOptions = processorOptions,
            ))
  }

  private fun List<File>.asSources(): List<Source> {
    return mapNotNull { file ->
      val relPath = file.relativeTo(File("../tests/src/main/java")).toString()
      val qName = relPath.substringBeforeLast(".").replace('/', '.')
      return@mapNotNull when (file.extension) {
        "java" -> Source.loadJavaSource(file, qName)
        "kt" -> Source.loadKotlinSource(file, relPath)
        else -> null
      }
    }
  }

  private fun getFiles(dir: File): List<File> {
    return dir.walkTopDown()
        .filter {
          !it.isDirectory && it.extension in setOf("kt", "java") && it.name != "ScopeImpl.java"
        }
        .toList()
  }

  private fun createProcessors(): Pair<Processor?, MotifSymbolProcessorProvider?> {
    val xProcConfig = XProcessingEnvConfig.DEFAULT.copy(false, true)
    val annotationProcessor =
        when (processorType) {
          ProcessorType.AP -> Processor()
          ProcessorType.KSP -> null
        }
    val symbolProcessorProvider =
        when (processorType) {
          ProcessorType.AP -> null
          ProcessorType.KSP -> MotifSymbolProcessorProvider(xProcConfig)
        }
    return annotationProcessor to symbolProcessorProvider
  }

  private fun assertSucceeded(result: TestCompilationResult) {
    if (!result.success || result.diagnostics[Diagnostic.Kind.ERROR].orEmpty().isNotEmpty()) {
      Truth.assertWithMessage(result.messages()).fail()
    }
  }

  private fun assertFailed(result: TestCompilationResult) {
    if (result.success) {
      Truth.assertWithMessage("Expected compilation to fail but encountered no errors.").fail()
    }
  }

  @Throws(Throwable::class)
  private fun runErrorTest(result: TestCompilationResult, graph: ResolvedGraph) {
    assertFailed(result)
    val expectedErrorString = getExistingErrorString()
    val actualErrorString = getActualErrorString(result)
    if (expectedErrorString != actualErrorString) {
      BufferedWriter(FileWriter(errorFile)).use { out -> out.write(actualErrorString) }
      Truth.assertWithMessage(
              """
        Error message has changed. The ERROR.txt file has been automatically updated by this test:
          1. Verify that the changes are correct.
          2. Commit the changes to source control.
        """.trimIndent())
          .that(actualErrorString)
          .isEqualTo(expectedErrorString)
    }
    runGraphTest(graph)
  }

  @Throws(Throwable::class)
  private fun runSuccessTest(testClass: Class<*>, graph: ResolvedGraph) {
    try {
      testClass.getMethod("run").invoke(null)
    } catch (e: InvocationTargetException) {
      throw e.cause!!
    }
    runGraphTest(graph)
  }

  @Throws(Throwable::class)
  private fun runGraphTest(graph: ResolvedGraph) {
    val expectedGraphString = getExistingGraphString()
    val actualGraphString = getActualGraphString(graph)
    if (expectedGraphString != actualGraphString) {
      BufferedWriter(FileWriter(graphFile)).use { out -> out.write(actualGraphString) }
      Truth.assertWithMessage(
              """
        Graph representation has changed. The GRAPH.txt file has been automatically updated by this test:
          1. Verify that the changes are correct.
          2. Commit the changes to source control.
        """.trimIndent())
          .fail()
    }
  }

  private fun getActualGraphString(graph: ResolvedGraph): String {
    val message = render(graph)
    val header =
        """
      ########################################################################
      #                                                                      #
      # This file is auto-generated by running the Motif compiler tests and  #
      # serves a as validation of graph correctness. IntelliJ plugin tests   #
      # also rely on this file to ensure that the plugin graph understanding #
      # is equivalent to the compiler's.                                     #
      #                                                                      #
      # - Do not edit manually.                                              #
      # - Commit changes to source control.                                  #
      # - Since this file is autogenerated, code review changes carefully to #
      #   ensure correctness.                                                #
      #                                                                      #
      ########################################################################

      """.trimIndent()
    return "$header\n$message\n"
  }

  @Throws(IOException::class)
  private fun getExistingGraphString(): String {
    return if (graphFile.exists()) {
      com.google.common.io.Files.asCharSource(graphFile, Charset.defaultCharset()).read()
    } else {
      ""
    }
  }

  private fun getActualErrorString(result: TestCompilationResult): String {
    val message = getMessage(result)
    val header =
        """
      ########################################################################
      #                                                                      #
      # This file is auto-generated by running the Motif compiler tests and  #
      # serves both as validation of error correctness and as a record of    #
      # the current compiler error output.                                   #
      #                                                                      #
      # - Do not edit manually.                                              #
      # - Commit changes to source control.                                  #
      # - Since this file is autogenerated, code review changes carefully to #
      #   ensure correctness.                                                #
      #                                                                      #
      ########################################################################

      """.trimIndent()
    return "$header\n$message\n"
  }

  private fun getMessage(result: TestCompilationResult): String {
    val header = toCompilerMessage(header).trimIndent()
    val footer = toCompilerMessage(footer).trimIndent()
    val resultMessage =
        result.messages().trimIndent().substringAfter(header).substringBefore(footer)
    return "$header$resultMessage$footer".prependIndent("  ")
  }

  private fun toCompilerMessage(message: String): String {
    return message.trim().prependIndent("  ")
  }

  @Throws(IOException::class)
  private fun getExistingErrorString(): String {
    return if (errorFile.exists()) {
      com.google.common.io.Files.asCharSource(errorFile, Charset.defaultCharset()).read()
    } else {
      ""
    }
  }
}

internal fun File.listFilesRecursively() = walkTopDown().filter { it.isFile }.toList()

private fun TestCompilationResult.messages(): String {
  return diagnostics[Diagnostic.Kind.ERROR].orEmpty().flatMap { it.msg.lines() }.joinToString(
      separator = "\n") { "  $it" }
}
